src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java

```diff
@@ -395,6 +401,23 @@ protected SCMProbe createProbe(SCMHead head, @CheckForNull SCMRevision revision)
         return new BitbucketSCMProbe(head, scmHelper.getFilePathClient(getProjectKey(), getRepositorySlug()));
     }
 
+    /**
+     * This method gets invoked by XStream after the {@link BitbucketSCMSource} object is unmarshalled.
+     */
+    @SuppressWarnings("unused") // Used by XStream
+    protected Object readResolve() {
+        if (traits != null) {
+            // Convert any legacy traits into their new equivalents
+            List<SCMSourceTrait> traits = this.traits.stream().map(BitbucketLegacyTraitConverter::maybeConvert)
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.toList());
+            this.traits.clear();
```

Comment:
As discussed one option is to change this to be a not final variable and just re-assign it.